ghost/milestone-emails/lib/Milestone.js

```diff
@@ -0,0 +1,229 @@
+        };
+    }
+
+    /** @private */
+    constructor(data) {
+        this.#id = data.id;
+        this.#name = data.name;
+        this.#type = data.type;
+        this.#value = data.value;
+        this.#currency = data.currency;
+        this.#createdAt = data.createdAt;
+        this.#emailSentAt = data.emailSentAt;
+    }
+
+    /**
+     * @param {any} data
+     * @returns {Promise<Milestone>}
+     */
+    static async create(data) {
+        // order of validation matters!
+        const id = validateId(data.id);
+        const type = validateType(data.type);
+        const currency = validateCurrency(type, data?.currency);
+        const value = validateValue(data.value);
+        const name = validateName(data.name, value, type, currency);
+        const emailSentAt = validateEmailSentAt(data);
+
+        /** @type Date */
+        let createdAt;
+        if (data.createdAt instanceof Date) {
+            createdAt = data.createdAt;
+        } else if (data.createdAt) {
+            createdAt = new Date(data.createdAt);
+            if (isNaN(createdAt.valueOf())) {
+                throw new ValidationError({
+                    message: 'Invalid Date'
+                });
+            }
+        } else {
+            createdAt = new Date();
+        }
+
+        return new Milestone({
+            id,
+            name,
+            type,
+            value,
+            currency,
+            createdAt,
+            emailSentAt
+        });
+    }
+};
+
+/**
+ *
+ * @param {ObjectID|string|null} id
+ *
+ * @returns {ObjectID}
+ */
+function validateId(id) {
+    if (!id) {
+        return new ObjectID();
+    }
+    if (typeof id === 'string') {
+        return ObjectID.createFromHexString(id);
+    }
+    if (id instanceof ObjectID) {
+        return id;
+    }
+    return new ObjectID();
+}
+
+/**
+ *
+ * @param {number|null} value
+ *
+ * @returns {number}
+ */
+function validateValue(value) {
+    if (!value || typeof value !== 'number' || value === 0) {
+        throw new ValidationError({
+            message: 'Invalid value'
+        });
+    }
+
+    return value;
+}
+
+/**
+ *
+ * @param {'arr'|'members'} type
+ *
+ * @returns {string}
+ */
+function validateType(type) {
+    if (type === 'arr') {
+        return 'arr';
+    }
+
+    return 'members';
+}
+
+/**
+ *
+ * @param {'arr'|'members'} type
+ * @param {string|null} currency
+ *
+ * @returns {string}
+ */
+function validateCurrency(type, currency) {
+    if (type === 'members') {
+        return null;
+    }
+
+    if (!currency || (currency && typeof currency !== 'string')) {
+        return 'usd';
+    }
+
+    if (currency.length > 3) {
+        return currency.slice(0, 3);
+    }
+
+    return currency;
+}
+
+/**
+ *
+ * @param {string} name
+ * @param {number} value
+ * @param {'arr'|'members'} type
+ * @param {string|null} currency
+ *
+ * @returns {string}
+ */
+function validateName(name, value, type, currency) {
```

Comment:
I don't think you need to store the `name` as `this.#name` it seems to be calculated entirely from other values so you can just add a getter for it like:
```
get name() {
  if (this.type === 'arr') {
    return `arr-${this.value}-${this.currency}`;
  }
  return `members-${this.value}`;
}
```